### PR TITLE
Add 2% treasury reserve liquidation fee on top of the one configured for assets.

### PR DESCRIPTION
**Pull request by lastperson**
Will need to deploy new collateral manager contract, then switch it address in the address provider.

### CODE REVIEW

**Review by x-moola — APPROVED** [review]

**Comment by x-moola** [comment]
Looking great! Do we also need to redeploy mToken contracts after adding  `function RESERVE_TREASURY_ADDRESS` to `contracts/interfaces/IAToken.sol`? 🤔

File: contracts/protocol/lendingpool/LendingPoolCollateralManagerWithReserve.sol [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+        user,
+        vars.collateralAtoken.RESERVE_TREASURY_ADDRESS(),
+        vars.collateralReserveAmount
+      );
+    }
+
+    // If the collateral being liquidated is equal to the user balance,
+    // we set the currency as not being used as collateral anymore
+    if (vars.maxCollateralToLiquidate == vars.userCollateralBalance) {
+      userConfig.setUsingAsCollateral(collateralReserve.id, false);
```
Comment (diwu1989) [inline contracts/protocol/lendingpool/LendingPoolCollateralManagerWithReserve.sol:242]: be careful with this one here, this is the code path that got Agave drained on a re-entrant token.
Let's think a bit more about how to make this safer.

The agave got hacked was:
- collateral has re-entrancy token
- hacker deposit collateral, borrow, and then call liquidate on himself
- liquidate transfer re-entrant token to hacker
- hacker flash deposit more collateral
- hacker borrow another token
- liquidate callback complete, mark isUsingCollateral as false
- hacker withdraw collateral

rinse and repeat against every agave market

**Review by diwu1989 — COMMENTED** [review]
Consider fixing liquidate call to prevent Agave style hack

**Comment by diwu1989** [comment]
image
https://www.aliens.com/livenews/latest/blocksecalert-agave-contract-hacked-hackers-make-around-dollar54-million-in-profits

File: contracts/protocol/lendingpool/LendingPoolCollateralManagerWithReserve.sol (hunk 1 above)
Comment (lastperson) [reply contracts/protocol/lendingpool/LendingPoolCollateralManagerWithReserve.sol:242]: Aave is specifically does not support re-entrant tokens.
Fixing it in this one place will not guarantee that re-entrant tokens will become safe to use.
We need to make sure that the added tokens are not re-entrant.

**Comment by lastperson** [comment]
> Looking great! Do we also need to redeploy mToken contracts after adding `function RESERVE_TREASURY_ADDRESS` to `contracts/interfaces/IAToken.sol`? thinking

I only added it to the interface, while it was already present in the implementation, so no redeployment is needed.

**Comment by lastperson** [comment]
@diwu1989 @x-moola added an Agave fix. And made an explainer on the vulnerability:
Idea is to deposit, borrow and then force the protocol to setUsingAsCollateral(false), so that attacker could withdraw the whole deposit without repaying first. Liquidation was only needed because it would do setUsingAsCollateral(false) after the unsafe call to the user.

File: contracts/protocol/lendingpool/LendingPoolCollateralManagerWithReserve.sol (hunk 1 above)
Comment (diwu1989) [reply contracts/protocol/lendingpool/LendingPoolCollateralManagerWithReserve.sol]: Yes, taking the Agave fix is a good idea, thank you!

**Comment by diwu1989** [comment]
LGTM with the agave fix, good job finding that.

**Pull request merged**